### PR TITLE
Implement basic element handles

### PR DESCRIPTION
**Pull request by aslushnikov**
This patch implements basic element handles which a backed with remote objects.

References #111

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -1225,3 +1275,4 @@ Contains `POST` data for `POST` requests.
 [Map]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map "Map"
 [selector]: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Selectors "selector"
 [Tracing]: https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#class-tracing "Tracing"
+[Element]: https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#class-element "Element"
```
Comment (JoelEinbinder) [inline docs/api.md]: [ElementHandle], and above.

File: lib/Connection.js [hunk 2]
```diff
@@ -167,6 +167,8 @@ class Session extends EventEmitter {
    * @return {!Promise<?Object>}
    */
   async send(method, params = {}) {
+    if (!this._connection)
+      return Promise.reject(new Error(`Protocol error (${method}): Session closed. Most likely the inspected page has been closed.`));
```
Comment (JoelEinbinder) [inline lib/Connection.js]: nit: inspected page seems like a devtools word.

File: lib/ElementHandle.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+   */
+  async release() {
+    if (this._released)
+      return;
+    this._released = true;
+    await helper.releaseObject(this._client, this._remoteObject);
+  }
+
+  /**
+   * @param {function()|string} pageFunction
```
Comment (JoelEinbinder) [inline lib/ElementHandle.js]: Don't allow a string if you don't allow a string.

File: lib/FrameManager.js [hunk 4]
```diff
@@ -171,12 +172,34 @@ class Frame {
    * @return {!Promise<(!Object|undefined)>}
    */
   async evaluate(pageFunction, ...args) {
+    let remoteObject = await this._rawEvaluate(pageFunction, ...args);
+    return await helper.serializeRemoteObject(this._client, remoteObject);
```
Comment (JoelEinbinder) [inline lib/FrameManager.js:176]: Let's make serializeRemoteObject return an ElementHandle if it is a node.

File: lib/Page.js [hunk 5]
```diff
@@ -551,23 +559,32 @@ class Page extends EventEmitter {
    * @return {!Promise}
    */
   async click(selector, options) {
-    await this.mainFrame().click(selector, options);
+    let handle = await this.$(selector);
+    console.assert(handle, 'No node found for selector: ' + selector);
+    await handle.click(options);
+    await handle.release();
```
Comment (JoelEinbinder) [inline lib/Page.js:565]: Do we need to await the release?

File: docs/api.md (hunk 1 above)
Comment (aslushnikov) [reply docs/api.md]: Done.

File: lib/Connection.js (hunk 2 above)
Comment (aslushnikov) [reply lib/Connection.js]: Done.

File: lib/ElementHandle.js (hunk 3 above)
Comment (aslushnikov) [reply lib/ElementHandle.js]: Done.

File: lib/Page.js (hunk 5 above)
Comment (aslushnikov) [reply lib/Page.js:565]: Yes, otherwise the following might throw:
```js
await page.click();
page.close();
```

Comment (JoelEinbinder) [reply lib/Page.js:565]: Shouldn't close/navigate mark all handles as released?

File: lib/FrameManager.js (hunk 4 above)
Comment (aslushnikov) [reply lib/FrameManager.js:176]: Let me do this in a follow-up. This implies the `console` event changes and possibly something else, might be some work to do.
Also, the `page.waitFor` should resolve to elementHandles as well (another follow-up)

File: lib/Page.js (hunk 5 above)
Comment (JoelEinbinder) [reply lib/Page.js:565]: In helper we just do a try-catch:
https://github.com/GoogleChrome/puppeteer/blob/master/lib/helper.js#L103

Comment (aslushnikov) [reply lib/Page.js:565]: It's the execution context being destroyed that renders all the element handles useless. (we don't release them explicitly for now)

It's nice to await `handle.release` so that there would be no in-flight messages when the connection is getting closed. (Connection rejects all the in-flight messages when it closes).

**Comment by aslushnikov** [comment]
@JoelEinbinder could you please re-review this (I had to force-push)

File: docs/api.md [hunk 6]
```diff
@@ -1055,6 +1054,66 @@ puppeteer.launch().then(async browser => {
...
+#### elementHandle.release()
+- returns: <[Promise]> Promise which resolves when the element handle is successfully released.
+
+The `elementHandle.release` method stops referencing the element handle.
+
+```js
+const {Browser} = require('puppeteer');
+const browser = new Browser();
+browser.newPage().then(async page =>
+  await page.setContent('<div>hello</div>');
```
Comment (JoelEinbinder) [inline docs/api.md:1108]: How is this code relevant to `element.release`?

**Review by JoelEinbinder — APPROVED** [review]
